File: INLThemes/INLThemes/Classes/INLThemeService.m
//
//  INLThemeProvider.m
//
//  Created by Tomas Hakel on 28/01/2016.
//  Copyright © 2016 Inloop. All rights reserved.
//

#import "INLThemeService.h"
#import "INLAssetService.h"
#import "UIViewController+INLThemeConsumer.h"


@interface INLThemeService ()

@property (strong, nonatomic) INLTheme * currentTheme;
@property (strong, nonatomic) NSMutableArray<id<INLThemeConsumer>> * themeConsumers;

@end


@implementation INLThemeService

+(instancetype)sharedService {
	static INLThemeService * sharedService = nil;
	static dispatch_once_t onceToken;
	dispatch_once(&onceToken, ^{
		sharedService = [[INLThemeService alloc] init];
		sharedService.themeConsumers = [@[] mutableCopy];
	});
	return sharedService;
}

#pragma mark - Register theme consumer

-(void)registerThemeConsumer:(id<INLThemeConsumer>)themeConsumer {
	[self.themeConsumers addObject:themeConsumer];
	[self applyTheme:self.currentTheme to:themeConsumer];
}

-(void)removeThemeConsumer:(id<INLThemeConsumer>)themeConsumer {
	[self.themeConsumers removeObject:themeConsumer];
}

#pragma mark - Apply theme

-(void)applyTheme:(INLTheme *)theme {
	self.currentTheme = theme;
	for (id<INLThemeConsumer> themeConsumer in self.themeConsumers) {
		[self applyTheme:theme to:themeConsumer];
	}
}

-(void)applyTheme:(INLTheme *)theme to:(id<INLThemeConsumer>)themeConsumer {
	[self applyTheme:theme toViews:[themeConsumer themedViews]];
}

-(void)applyTheme:(INLTheme *)theme toViews:(NSArray<id<INLThemedView>> *)themedViews {
	for (id<INLThemedView> view in themedViews) {
		INLThemeElement * themeElement = theme.uiElements[view.elementId];
		if (themeElement) {
			[view applyTheme:themeElement];
		}
	}
}

#pragma mark - Convinience

+(void)registerThemeConsumer:(id<INLThemeConsumer>)themeConsumer {
	[[INLThemeService sharedService] registerThemeConsumer:themeConsumer];
}

+(void)removeThemeConsumer:(id<INLThemeConsumer>)themeConsumer {
	[[INLThemeService sharedService] removeThemeConsumer:themeConsumer];
}

+(void)applyTheme:(INLTheme *)theme {
	[[INLThemeService sharedService] applyTheme:theme];
}

+(void)applyTheme:(INLTheme *)theme to:(id<INLThemeConsumer>)themeConsumer {
	[[INLThemeService sharedService] applyTheme:theme to:themeConsumer];
}

@end

File: INLThemes/INLThemes/Classes/INLThemeService.swift
//
//  INLThemeService.swift
//  INLThemes
//
//  Created by Tomas Hakel on 01/03/2016.
//  Copyright © 2016 Inloop. All rights reserved.
//

import Foundation


@objc(INLThemeService)
open class ThemeService: NSObject {
    static let shared = ThemeService()

	var currentTheme: Theme?
	var themeConsumers = [ThemeConsumer]()

	// MARK: - Register theme consumer
	open func register(_ themeConsumer: ThemeConsumer) {
		themeConsumers.append(themeConsumer)
		if let theme = currentTheme {
			apply(theme, to: themeConsumer)
		}
	}

	open func remove(_ themeConsumer: ThemeConsumer) {
		themeConsumers = themeConsumers.filter { !$0.isEqual(themeConsumer) }
	}

	// MARK: - Apply theme
	open func apply(_ theme: Theme) {
		currentTheme = theme
		themeConsumers.forEach { apply(theme, to: $0) }
	}

	open func apply(_ theme: Theme, to themeConsumer: ThemeConsumer) {
		apply(theme, toViews: themeConsumer.themedViews())
	}

	open func applyTheme(to themeConsumer: ThemeConsumer) {
		if let currentTheme = self.currentTheme {
			apply(currentTheme, to: themeConsumer)
		}
	}

	func apply(_ theme: Theme, toViews themedViews: [ThemedView]) {
		themedViews.forEach {
			if let element = theme.uiElements[$0.elementId] {
				$0.apply(element)
			}
		}
	}

	// MARK: - Convinience
	open class func register(_ themeConsumer: ThemeConsumer) {
        shared.register(themeConsumer)
	}

	open class func remove(_ themeConsumer: ThemeConsumer) {
        shared.remove(themeConsumer)
	}

	open class func apply(_ theme: Theme) {
		shared.apply(theme)
	}

	open class func apply(_ theme: Theme, to themeConsumer: ThemeConsumer) {
		shared.apply(theme, to: themeConsumer)
	}

	open class func applyTheme(to themeConsumer: ThemeConsumer) {
        shared.applyTheme(to: themeConsumer)
	}
}
